RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc

```diff
@@ -0,0 +1,631 @@
+	useTriple=true;
+      }
+    }
+    else {
+      doFit(reconstructedVals,0);
+      useTriple=true;
+    }
+  }
+  else{
+    reconstructedVals.at(0) = 0.; //energy
+    reconstructedVals.at(1) = -9999.; //time
+    reconstructedVals.at(2) = -9999.; //chi2
+  }
+  
+  reconstructedEnergy = reconstructedVals[0]*channelData.tsGain(0);
+  reconstructedTime = reconstructedVals[1];
+  chi2 = reconstructedVals[2];
+
+}
+
+void MahiFit::doFit(std::array<float,3> &correctedOutput, int nbx) const {
+
+  unsigned int bxSize=1;
+
+  if (nbx==1) {
+    nnlsWork_.bxOffset = 0;
+  }
+  else {
+    bxSize = bxSizeConf_;
+    nnlsWork_.bxOffset = bxOffsetConf_;
+  }
+
+  nnlsWork_.bxs.resize(bxSize);
+
+  if (nbx==1) {
+    nnlsWork_.bxs.coeffRef(0) = 0;
+  }
+  else {
+    for (unsigned int iBX=0; iBX<bxSize; iBX++) {
+      nnlsWork_.bxs.coeffRef(iBX) = activeBXs_[iBX];
+    }
+  }
+
+  nnlsWork_.nPulseTot = bxSize;
+
+  if (dynamicPed_) {
+    nnlsWork_.nPulseTot++;
+    nnlsWork_.bxs.resize(nnlsWork_.nPulseTot);
+    nnlsWork_.bxs[nnlsWork_.nPulseTot-1] = pedestalBX_;
+  }
+
+  nnlsWork_.pulseMat.resize(nnlsWork_.tsSize,nnlsWork_.nPulseTot);
+  nnlsWork_.ampVec = PulseVector::Zero(nnlsWork_.nPulseTot);
+  nnlsWork_.errVec = PulseVector::Zero(nnlsWork_.nPulseTot);
+
+  int offset=0;
+  for (unsigned int iBX=0; iBX<nnlsWork_.nPulseTot; iBX++) {
+    offset=nnlsWork_.bxs.coeff(iBX);
+
+    nnlsWork_.pulseShapeArray[iBX] = FullSampleVector::Zero(MaxFSVSize);
+    nnlsWork_.pulseDerivArray[iBX] = FullSampleVector::Zero(MaxFSVSize);
+    nnlsWork_.pulseCovArray[iBX]   = FullSampleMatrix::Constant(0);
+
+    if (offset==pedestalBX_) {
+      nnlsWork_.ampVec.coeffRef(iBX) = sqrt(nnlsWork_.pedConstraint.coeff(0));
+    }
+    else {
+
+      updatePulseShape(nnlsWork_.amplitudes.coeff(nnlsWork_.tsOffset + offset), 
+		       nnlsWork_.pulseShapeArray[iBX], 
+		       nnlsWork_.pulseDerivArray[iBX],
+		       nnlsWork_.pulseCovArray[iBX]);
+      
+      nnlsWork_.ampVec.coeffRef(iBX)=0;
+
+      nnlsWork_.pulseMat.col(iBX) = nnlsWork_.pulseShapeArray[iBX].segment(nnlsWork_.fullTSOffset - offset, nnlsWork_.tsSize);
+
+    }
+  }
+
+  nnlsWork_.pulseMat.col(nnlsWork_.nPulseTot-1) = SampleVector::Ones(nnlsWork_.tsSize);
+
+  nnlsWork_.aTaMat.resize(nnlsWork_.nPulseTot, nnlsWork_.nPulseTot);
+  nnlsWork_.aTbVec.resize(nnlsWork_.nPulseTot);
+
+  double chiSq = minimize(); 
+
+  nnlsWork_.residuals = nnlsWork_.pulseMat*nnlsWork_.ampVec - nnlsWork_.amplitudes;
+
+  bool foundintime = false;
+  unsigned int ipulseintime = 0;
+
+  for (unsigned int iBX=0; iBX<nnlsWork_.nPulseTot; ++iBX) {
+    if (nnlsWork_.bxs.coeff(iBX)==0) {
+      ipulseintime = iBX;
+      foundintime = true;
+    }
+  }
+
+  if (foundintime) {
+    correctedOutput.at(0) = nnlsWork_.ampVec.coeff(ipulseintime); //charge
+    double arrivalTime = calculateArrivalTime();
+    correctedOutput.at(1) = arrivalTime; //time
+    correctedOutput.at(2) = chiSq; //chi2
+
+  }
```

Comment:
shouldn't there be a way to communicate that foundintime was false?

Reply:
Yes, I agree, & there are a number of other checks that need to be added -- ex. that the configured BX's include the in-time bunch crossing, that all the BX's are within range of the 8/10 TS... may I follow up on these sanity checks in a later PR?

Reply:
now that the initialization of the input arguments is in place, it's OK to sort it out later.